specfile/sources.py

```diff
@@ -0,0 +1,414 @@
+
+        Returns:
+            Extracted index or None if there isn't one.
+        """
+        tokens = re.split(r"(\d+)", self.tag.name, maxsplit=1)
+        if len(tokens) > 1:
+            return tokens[1]
+        return None
+
+    @property
+    def index(self) -> int:
+        """Numeric index of the source."""
+        return int(self._get_index())
+
+    @property
+    def index_digits(self) -> int:
+        """Number of digits the index is formed by."""
+        return len(self._get_index() or "")
+
+    @property
+    def url(self) -> str:
+        """Literal URL of the source as stored in the spec file."""
+        return self.tag.value
+
+    @url.setter
+    def url(self, value: str) -> None:
+        self.tag.value = value
+
+    @property
+    def expanded_url(self) -> str:
+        """URL of the source after expanding macros."""
+        return self.tag.expanded_value
+
+    @property
+    def comments(self) -> Comments:
+        """List of comments associated with the source."""
+        return self.tag.comments
+
+
+class ListSource(GenericSource):
+    """
+    Class that represents a source backed by a line in a %sourcelist/%patchlist section.
+
+    Attributes:
+        source: Source object that this source represents.
+    """
+
+    def __init__(self, source: Source, index: int) -> None:
+        self.source = source
+        self._index = index
+
+    def __repr__(self) -> str:
+        source = repr(self.source)
+        return f"ListSource({source}, {self._index})"
+
+    @property
+    def index(self) -> int:
+        """Numeric index of the source."""
+        return self._index
+
+    @property
+    def url(self) -> str:
+        """Literal URL of the source as stored in the spec file."""
+        return self.source.url
+
+    @url.setter
+    def url(self, value: str) -> None:
+        self.source.url = value
+
+    @property
+    def expanded_url(self) -> str:
+        """URL of the source after expanding macros."""
+        return self.source.expanded_url
+
+    @property
+    def comments(self) -> Comments:
+        """List of comments associated with the source."""
+        return self.source.comments
+
+
+class Sources:
+    """
+    Class that represents a list of all sources.
+
+    Attributes:
+        tags: All spec file tags.
+        sourcelists: List of all %sourcelist sections.
+        allow_duplicates: Whether to allow duplicate entries when adding new sources.
+    """
+
+    prefix = "Source"
+
+    def __init__(
+        self, tags: Tags, sourcelists: List[SourceList], allow_duplicates: bool = False
+    ) -> None:
+        self.tags = tags
+        self.sourcelists = sourcelists
+        self.allow_duplicates = allow_duplicates
+
+    def __repr__(self) -> str:
+        tags = repr(self.tags)
+        sourcelists = repr(self.sourcelists)
+        allow_duplicates = repr(self.allow_duplicates)
+        return f"{self.__class__.__name__}({tags}, {sourcelists}, {allow_duplicates})"
```

Comment:
So that `Patches` doesn't have to reimplement `__repr__()`. Do you think it would be better?

Reply:
Thanks for the explanation, I didn't realize it's due to inheritance. I think this way it's fine, it's following DRY principle :) But maybe a comment there to avoid confusion would be useful. If someone new came to this project (or we are changing something in a few years :D ) and didn't notice inheritance, they could change it for "consistency" with other classes but that would cause a bug that is hard to notice (repr isn't tested anywhere). So I'd rather put a comment there to prevent it.